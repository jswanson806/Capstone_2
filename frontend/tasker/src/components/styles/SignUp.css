.signup-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    padding-bottom: 30%;
    background-color: #f2f2f2;
  }
  
  .signup-centered-card {
    width: 300px; 
    padding: 20px;
    border-radius: 10px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.2); /* Add a drop shadow */
    background-color: #fff; /* Card background color */
  }
  
  .signup-centered-card h5 {
    text-align: center;
    margin-bottom: 20px; /* Add spacing below the title */
  }
  
  .signup-centered-card input {
    width: 100%;
    padding: 10px;
    margin-bottom: 10px;
    border: 1px solid #ccc;
    border-radius: 5px;
  
  }
  
  .signup-centered-card button.signup-button {
    width: 100%;
    padding: 10px;
    background-color: #007bff; /* Button background color */
    color: #fff; /* Button text color */
    border: none;
    border-radius: 5px;
    cursor: pointer;
  
  }
  
  