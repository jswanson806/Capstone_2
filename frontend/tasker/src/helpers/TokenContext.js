import React, { createContext, useState } from 'react';

// Create the context
const TokenContext = createContext();

// Create a provider for the context
const TokenProvider = ({ children }) => {
  const [token, setToken] = useState(localStorage.getItem('token') || '');

  // Function to update the token in the state and localStorage
  const updateToken = (newToken) => {
    setToken(newToken);
    localStorage.setItem('token', newToken);
  };

  return (
    <TokenContext.Provider value={{ token, updateToken }}>
      {children}
    </TokenContext.Provider>
  );
};

export { TokenContext, TokenProvider };
