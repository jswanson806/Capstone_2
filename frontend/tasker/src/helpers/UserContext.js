import React, { createContext, useState } from 'react';

// Create the context
const UserContext = createContext();

// Create a provider for the context
const UserProvider = ({ children }) => {
  const [user, setUser] = useState(localStorage.getItem('user') || '');

  // Function to update the user in the state and localStorage
  const updateUser = (newUser) => {
    setUser(newUser);
    localStorage.setItem('user', newUser);
  };

  return (
    <UserContext.Provider value={{ user, updateUser }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
