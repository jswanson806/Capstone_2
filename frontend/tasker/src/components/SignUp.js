import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import { TokenContext } from "../helpers/TokenContext";
import { UserContext } from "../helpers/UserContext";
import TaskerApi from "../api";
import {Form, Input, Label, FormGroup, Button, ButtonGroup, UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem} from "reactstrap";
import "./styles/SignUp.css";

/**
 * SignUp component for user registration.
 *
 * This component renders a sign-up form where users can enter their email, first name,
 * last name, phone number, and password to register for an account. Users can select either 'user' or 'worker' as the registration type.
 * Upon successful registration, the user is logged in, and their token is stored in the local storage
 * for further authentication. The user is then redirected to the dashboard.
 *
 * @returns {JSX.Element} The rendered sign-up form component.
 */


const SignUp = () => {

    const navigate = useNavigate();

    
    const INITIAL_STATE = {
        email: '',
        firstName: '',
        lastName: '',
        phone: '',
        password: '',
    }

    const [formData, setFormData] = useState(INITIAL_STATE);
    const [ isWorker, setIsWorker ] = useState(false);
    const { token, updateToken } = useContext(TokenContext);
    const { user, updateUser } = useContext(UserContext);

    // accepts a value and removes non-integer characters and inserts hyphens
    function formatPhoneNumber(value) {
        let cleanedValue = value.replace(/\D/g, '');
        let match = cleanedValue.match(/^(\d{0,3})(\d{0,3})(\d{0,4})$/);
        
        if (!match) return cleanedValue;
    
        return match[1] + (match[1] && match[2] ? '-' : '') + match[2] + (match[3] ? '-' : '') + match[3];
      }


    const handleChange = (e) => {
        const {name, value} = e.target;

        // conditionally format field for phone number or else just update the field
        if(name === "phone"){
            let inputValue = e.target.value.replace(/\D/g, ''); // Remove non-digit characters
            let formattedValue = formatPhoneNumber(inputValue);
            
            // update field with formatted value as user types
            e.target.value = formattedValue;

            setFormData(formData => ({
                ...formData,
                [name]: formattedValue
            }))

        } else {

            setFormData(formData => ({
                ...formData,
                [name]: value
            }))
        }
    }

    const handleSubmit = async (e) => {
        e.preventDefault();

        let userInfo = {user: {
            email: formData.email,
            firstName: formData.firstName,
            lastName: formData.lastName,
            phone: formData.phone,
            password: formData.password,
            isWorker: false,
        }};

        

        // set isWorker property on userInfo object based on state of isWorker
        if(isWorker){
            userInfo.user.isWorker = true;
        }

        try {
            // register call to api
            const registerToken = await TaskerApi.registerUser(userInfo);
        
            // check for registerToken to be true
            if(registerToken) {
                // login call to api
                const res = await TaskerApi.login(userInfo.user.email, userInfo.user.password);
                const { token, user } = res;
                const newUser = JSON.stringify({id: user.id, email: user.email, isWorker: user.isWorker});
                // set token in local storage for further auth
                updateToken(token);
                updateUser(newUser);
            }

            // clear the form data
            setFormData(INITIAL_STATE);
            // redirect to dashboard
            navigate("/dashboard");

        } catch(err) {
            return err;
        }
        
    }

    return (
        <div className="signup-container">
            <div className="signup-centered-card">

                {/* render title depending on state of isWorker */}
                <h5>{!isWorker ? 'Sign Up to Post Jobs' : 'Sign Up to Find Jobs'}</h5>

                {/* render button to toggle between user and worker signup */}
                <ButtonGroup className="my-2">
                  <UncontrolledDropdown>
                    <DropdownToggle caret>
                      {isWorker ? <>Worker</> : <>User</>}
                    </DropdownToggle>
                    <DropdownMenu>
                      <DropdownItem onClick={() => setIsWorker(true)}>
                        Worker
                      </DropdownItem>
                      <DropdownItem onClick={() => setIsWorker(false)}>
                        User
                      </DropdownItem>
                    </DropdownMenu>
                  </UncontrolledDropdown>
                </ButtonGroup>
                
                {/* form for registering */}
                <Form onSubmit={handleSubmit}>
                    <FormGroup>
                    <Label htmlFor="email">Email</Label>
                    <Input 
                        id="email"
                        type="text"
                        name="email"
                        placeholder="Email Address"
                        data-testid="signup-form-email-input"
                        value={FormData.email}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <FormGroup>
                    <Label htmlFor="firstName">First Name</Label>
                    <Input
                        id="firstName"
                        type="text"
                        name="firstName"
                        placeholder="First Name"
                        data-testid="signup-form-first-name-input"
                        value={FormData.firstName}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <FormGroup>
                    <Label htmlFor="lastName">Last Name</Label>
                    <Input
                        id="lastName"
                        type="text"
                        name="lastName"
                        placeholder="Last Name"
                        data-testid="signup-form-last-name-input"
                        value={FormData.lastName}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <FormGroup>
                    <Label htmlFor="phone">Phone</Label>
                    <Input
                        id="phone"
                        type="tel"
                        name="phone"
                        placeholder="Phone"
                        data-testid="signup-form-phone-input"
                        value={FormData.phone}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <FormGroup> 
                    <Label htmlFor="password">Password</Label>
                    <Input
                        id="password"
                        type="password"
                        name="password"
                        placeholder=""
                        data-testid="signup-form-password-input"
                        value={FormData.password}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <Button className="signup-button" color="info" type="submit" data-testid="signup-form-button">Signup</Button>
                </Form>
                
            </div>
        </div>
    );
}

export default SignUp;