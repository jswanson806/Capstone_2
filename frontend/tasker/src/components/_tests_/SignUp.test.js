import React from 'react';
import SignUp from '../SignUp.js';
import {act, render, screen, fireEvent} from '@testing-library/react';
import { MemoryRouter } from "react-router-dom";
import { TokenContext } from '../../helpers/TokenContext.js';
import TaskerApi from '../../api.js';
import { UserContext } from '../../helpers/UserContext.js';

const userValue = { user: '{"id": 1, "email": "test@email.com", "isWorker": true}' };

describe("signup form smoke and snapshot tests", () => {

    it("should render without crashing", async () => {

        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <SignUp />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });
    });

    it("should match snapshot", async() => {
        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <SignUp />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    })
})

describe("Handles form input correctly", () => {

    it('updates form input correctly', async () => {
        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <SignUp />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });

        const emailInput = screen.getByTestId("signup-form-email-input");
        const firstName = screen.getByTestId("signup-form-first-name-input");
        const lastName = screen.getByTestId("signup-form-last-name-input");
        const phone = screen.getByTestId("signup-form-phone-input");
        const password = screen.getByTestId("signup-form-password-input");
        fireEvent.change(emailInput, {target: {value: "test@email.com"}});
        fireEvent.change(firstName, {target: {value: "Testy"}});
        fireEvent.change(lastName, {target: {value: "McTesty"}});
        fireEvent.change(phone, {target: {value: "444-444-4444"}});
        fireEvent.change(password, {target: {value: "Password1"}});
        expect(emailInput).toHaveValue("test@email.com");
        expect(firstName).toHaveValue("Testy");
        expect(lastName).toHaveValue("McTesty");
        expect(phone).toHaveValue("444-444-4444");
        expect(password).toHaveValue("Password1");

    })

    it('should handle form submission', async () => {

        // mock the registerUser and login functions of the api
        const mockRegisterUser = jest.spyOn(TaskerApi, 'registerUser').mockResolvedValue('dummy token');

        const mockLogin = jest.spyOn(TaskerApi, 'login')

        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <SignUp />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });

        const emailInput = screen.getByTestId("signup-form-email-input");
        const firstName = screen.getByTestId("signup-form-first-name-input");
        const lastName = screen.getByTestId("signup-form-last-name-input");
        const phone = screen.getByTestId("signup-form-phone-input");
        const password = screen.getByTestId("signup-form-password-input");
        fireEvent.change(emailInput, {target: {value: "test@email.com"}});
        fireEvent.change(firstName, {target: {value: "Testy"}});
        fireEvent.change(lastName, {target: {value: "McTesty"}});
        fireEvent.change(phone, {target: {value: "444-444-4444"}});
        fireEvent.change(password, {target: {value: "Password1"}});

        const btn = screen.getByTestId("signup-form-button");

        await act(async () => {
            fireEvent.click(btn);
        })

        expect(mockRegisterUser).toHaveBeenCalledWith({user: { 
            email: "test@email.com", 
            first_name: "Testy",
            last_name: "McTesty",
            is_worker: false,
            phone: "444-444-4444",
            password: "Password1"
        }});

        expect(mockLogin).toHaveBeenCalledWith("test@email.com", "Password1")
    })

})