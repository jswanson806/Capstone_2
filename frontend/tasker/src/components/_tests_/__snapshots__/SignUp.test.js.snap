// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup form smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="signup-container"
  >
    <div
      class="signup-centered-card"
    >
      <h5>
        Sign Up to Post Jobs
      </h5>
      <div
        class="my-2 btn-group"
        role="group"
      >
        <div
          class="dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            type="button"
          >
            User
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Worker
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              User
            </button>
          </div>
        </div>
      </div>
      <form
        class=""
      >
        <div
          class="form-group"
        >
          <label
            class=""
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            data-testid="signup-form-email-input"
            id="email"
            name="email"
            placeholder="Email Address"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            data-testid="signup-form-first-name-input"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            data-testid="signup-form-last-name-input"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="phone"
          >
            Phone
          </label>
          <input
            class="form-control"
            data-testid="signup-form-phone-input"
            id="phone"
            name="phone"
            placeholder="Phone"
            type="tel"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="signup-form-password-input"
            id="password"
            name="password"
            placeholder=""
            type="password"
            value=""
          />
        </div>
        <button
          class="signup-button btn btn-info"
          data-testid="signup-form-button"
          type="submit"
        >
          Signup
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
